daemon/containerd/image_list.go

```diff
@@ -43,21 +91,90 @@ func (i *ImageService) Images(ctx context.Context, opts types.ImageListOptions)
 		})
 	}
 
+	if opts.SharedSize {
+		for n, chainIDs := range root {
+			sharedSize, err := computeSharedSize(*chainIDs, layers, snapshotSizeFn)
+			if err != nil {
+				return nil, err
+			}
+			ret[n].SharedSize = sharedSize
+		}
+	}
+
 	return ret, nil
 }
 
-func computeVirtualSize(ctx context.Context, image containerd.Image, snapshotter snapshots.Snapshotter) (int64, error) {
-	var virtualSize int64
-	diffIDs, err := image.RootFS(ctx)
+func (i *ImageService) setupFilters(ctx context.Context, opts types.ImageListOptions) ([]imageFilterFunc, error) {
+	var filters []imageFilterFunc
+	err := opts.Filters.WalkValues("before", func(value string) error {
+		ref, err := reference.ParseDockerRef(value)
+		if err != nil {
+			return err
+		}
+		img, err := i.client.GetImage(ctx, ref.String())
+		if img != nil {
+			t := img.Metadata().CreatedAt
+			filters = append(filters, func(image containerd.Image) bool {
+				created := image.Metadata().CreatedAt
+				return created.Equal(t) || created.After(t)
+			})
+		}
+		return err
+	})
 	if err != nil {
-		return virtualSize, err
+		return nil, err
+	}
+
+	err = opts.Filters.WalkValues("since", func(value string) error {
+		ref, err := reference.ParseDockerRef(value)
+		if err != nil {
+			return err
+		}
+		img, err := i.client.GetImage(ctx, ref.String())
+		if img != nil {
+			t := img.Metadata().CreatedAt
+			filters = append(filters, func(image containerd.Image) bool {
+				created := image.Metadata().CreatedAt
+				return created.Equal(t) || created.Before(t)
+			})
+		}
+		return err
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	if opts.Filters.Contains("label") {
+		filters = append(filters, func(image containerd.Image) bool {
+			return opts.Filters.MatchKVList("label", image.Labels())
+		})
 	}
-	for _, chainID := range identity.ChainIDs(diffIDs) {
-		usage, err := snapshotter.Usage(ctx, chainID.String())
+	return filters, nil
+}
+
+func computeVirtualSize(chainIDs []digest.Digest, sizeFn func(d digest.Digest) (int64, error)) (int64, error) {
```

Comment:
While working on upstreaming this patch, I noticed that `image.Usage()` also allows for snapshots to be included (does that include all layers?); https://github.com/moby/moby/blob/2fbc30739b437d91f7a444f49143305e5f0a303e/vendor/github.com/containerd/containerd/image.go#L210-L231

Which can be enabled through `WithSnapshotUsage()` https://github.com/moby/moby/blob/2fbc30739b437d91f7a444f49143305e5f0a303e/vendor/github.com/containerd/containerd/image.go#L92-L94

Perhaps we should also look into that? Although downside of that would be that we may miss the "per-snapshot" cache implemented in this patch (maybe we can introduce a `WithCache()` functional argument 🤔